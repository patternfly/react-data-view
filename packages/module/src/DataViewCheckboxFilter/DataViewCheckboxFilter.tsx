import React from 'react';
import {
  Badge,
  Menu,
  MenuContent,
  MenuItem,
  MenuList,
  MenuProps,
  MenuToggle,
  Popper,
  ToolbarChip,
  ToolbarFilter,
} from '@patternfly/react-core';
import { FilterIcon } from '@patternfly/react-icons';
import { DataViewFilterOption } from '../DataViewFilters';

const isToolbarChip = (chip: string | ToolbarChip): chip is ToolbarChip =>
  typeof chip === 'object' && 'key' in chip;

export const isDataViewFilterOption = (obj: unknown): obj is DataViewFilterOption =>
  !!obj &&
  typeof obj === 'object' &&
  'label' in obj &&
  'value' in obj &&
  typeof (obj as DataViewFilterOption).value === 'string';

/** extends MenuProps */
export interface DataViewCheckboxFilterProps extends Omit<MenuProps, 'onSelect' | 'onChange'> {
  /** Unique key for the filter attribute */
  filterId: string;
  /** Array of current filter values */
  value?: string[];
  /** Filter title displayed in the toolbar */
  title: string;
  /** Placeholder text of the menu */
  placeholder?: string;
  /** Filter options displayed */
  options: (DataViewFilterOption | string)[];
  /** Callback for updating when item selection changes. */
  onChange?: (event?: React.MouseEvent, values?: string[]) => void;
  /** Controls visibility of the filter in the toolbar */
  showToolbarItem?: boolean;
  /** Controls visibility of the filter icon */
  showIcon?: boolean;
  /** Controls visibility of the selected items badge */
  showBadge?: boolean;
  /** Custom OUIA ID */
  ouiaId?: string;
}

export const DataViewCheckboxFilter: React.FC<DataViewCheckboxFilterProps> = ({
  filterId,
  title,
  value = [],
  onChange,
  placeholder,
  options = [],
  showToolbarItem,
  showIcon = !placeholder,
  showBadge = !placeholder,
  ouiaId = 'DataViewCheckboxFilter',
  ...props
}: DataViewCheckboxFilterProps) => {
  const [ isOpen, setIsOpen ] = React.useState(false);
  const toggleRef = React.useRef<HTMLButtonElement>(null);
  const menuRef = React.useRef<HTMLDivElement>(null);
  const containerRef = React.useRef<HTMLDivElement>(null);

  const normalizeOptions = React.useMemo(
    () =>
      options.map(option =>
        typeof option === 'string'
          ? { label: option, value: option }
          : option
      ),
    [ options ]
  );

  const handleToggleClick = (event: React.MouseEvent) => {
    event.stopPropagation();
    setTimeout(() => {
      const firstElement = menuRef.current?.querySelector('li > button:not(:disabled)') as HTMLElement;
      firstElement?.focus();
    }, 0);
    setIsOpen(prev => !prev);
  };

  const handleSelect = (event?: React.MouseEvent, itemId?: string | number) => {
    const activeItem = String(itemId);
    const isSelected = value.includes(activeItem);

    onChange?.(
      event,
      isSelected ? value.filter(item => item !== activeItem) : [ activeItem, ...value ]
    );
  };

  const handleClickOutside = (event: MouseEvent) => 
    isOpen &&
    menuRef.current && toggleRef.current &&
    !menuRef.current.contains(event.target as Node) && !toggleRef.current.contains(event.target as Node)
    && setIsOpen(false);


  React.useEffect(() => {
    window.addEventListener('click', handleClickOutside);
    return () => {
      window.removeEventListener('click', handleClickOutside);
    };
  }, [ isOpen ]);  // eslint-disable-line react-hooks/exhaustive-deps

  return (
    <ToolbarFilter
      key={ouiaId}
      data-ouia-component-id={ouiaId}
      chips={value.map(item => {
        const activeOption = normalizeOptions.find(option => option.value === item);
        return ({ key: activeOption?.value as string, node: activeOption?.label })
      })}
      deleteChip={(_, chip) =>
        onChange?.(undefined, value.filter(item => item !== (isToolbarChip(chip) ? chip.key : chip)))
      }
      categoryName={title}
      showToolbarItem={showToolbarItem}
    >
      <Popper
        trigger={
          <MenuToggle
            ouiaId={`${ouiaId}-toggle`}
            ref={toggleRef}
            onClick={handleToggleClick}
            isExpanded={isOpen}
            icon={showIcon ? <FilterIcon /> : undefined}
            badge={value.length > 0 && showBadge ? <Badge data-ouia-component-id={`${ouiaId}-badge`} isRead>{value.length}</Badge> : undefined}
            style={{ width: '200px' }}
          >
            {placeholder ?? title}
          </MenuToggle>
        }
        triggerRef={toggleRef}
        popper={
          <Menu
            ref={menuRef}
            ouiaId={`${ouiaId}-menu`}
            onSelect={handleSelect}
            selected={value}
            {...props}
          >
            <MenuContent>
              <MenuList>
                {normalizeOptions.map(option => (
                  <MenuItem
                    data-ouia-component-id={`${ouiaId}-filter-item-${option.value}`}
                    key={option.value}
                    itemId={option.value}
                    isSelected={value.includes(option.value)}
                    hasCheckbox
                  >
                    {option.label}
                  </MenuItem>
                ))}
              </MenuList>
            </MenuContent>
          </Menu>
        }
        popperRef={menuRef}
        appendTo={containerRef.current || undefined}
        aria-label={`${title ?? filterId} filter`}
        isVisible={isOpen}
      />
    </ToolbarFilter>
  );
};

export default DataViewCheckboxFilter;
