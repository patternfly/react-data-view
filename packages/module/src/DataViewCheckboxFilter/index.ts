export { default } from './DataViewCheckboxFilter';
export * from './DataViewCheckboxFilter';
