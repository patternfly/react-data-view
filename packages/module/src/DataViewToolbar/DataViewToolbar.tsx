import React, { PropsWithChildren, useRef } from 'react';
import { Button, Toolbar, ToolbarContent, ToolbarItem, ToolbarItemVariant, ToolbarProps } from '@patternfly/react-core';

/** extends ToolbarProps */
export interface DataViewToolbarProps extends Omit<PropsWithChildren<ToolbarProps>, 'ref'> {
  /** Toolbar className */
  className?: string;
  /** Custom OUIA ID */
  ouiaId?: string;
  /** React node to display bulk select */
  bulkSelect?: React.ReactNode;
  /** React node to display pagination */
  pagination?: React.ReactNode;
  /** React node to display actions */
  actions?: React.ReactNode;
  /** React node to display filters */
  filters?: React.ReactNode;
  /** React node to display custom filter labels */
  customLabelGroupContent?: React.ReactNode;
}

export const DataViewToolbar: React.FC<DataViewToolbarProps> = ({ className, ouiaId = 'DataViewToolbar', bulkSelect, actions, pagination, filters, customLabelGroupContent, clearAllFilters, children, ...props }: DataViewToolbarProps) => {
  const defaultClearFilters = useRef(
    <ToolbarItem>
      <Button ouiaId={`${ouiaId}-clear-all-filters`} variant="link" onClick={clearAllFilters} isInline>
        Clear filters
      </Button>
    </ToolbarItem>
  );
  return (
    <Toolbar ouiaId={ouiaId} className={className} customLabelGroupContent={customLabelGroupContent ?? defaultClearFilters.current} {...props}>
      <ToolbarContent>
        {bulkSelect && (
          <ToolbarItem data-ouia-component-id={`${ouiaId}-bulk-select`}>
            {bulkSelect}
          </ToolbarItem>
        )}
        {filters && (
          <ToolbarItem>
            {filters}
          </ToolbarItem>
        )}
        {actions && (
          <ToolbarItem variant={ToolbarItemVariant['overflow-menu']}>
            {actions}
          </ToolbarItem>
        )}
        {pagination && (
          <ToolbarItem variant={ToolbarItemVariant.pagination} data-ouia-component-id={`${ouiaId}-pagination`}>
            {pagination}
          </ToolbarItem>
        )}
        {children}
      </ToolbarContent>
    </Toolbar>
  )
};

export default DataViewToolbar;

