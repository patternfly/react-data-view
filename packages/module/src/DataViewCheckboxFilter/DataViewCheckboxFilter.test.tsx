import React from 'react';
import { render } from '@testing-library/react';
import DataViewCheckboxFilter, { DataViewCheckboxFilterProps } from './DataViewCheckboxFilter';
import DataViewToolbar from '../DataViewToolbar';

describe('DataViewCheckboxFilter component', () => {
  const defaultProps: DataViewCheckboxFilterProps = {
    filterId: 'test-checkbox-filter',
    title: 'Test Checkbox Filter',
    value: [ 'workspace-one' ],
    options: [
      { label: 'Workspace one', value: 'workspace-one' },
      { label: 'Workspace two', value: 'workspace-two' },
      { label: 'Workspace three', value: 'workspace-three' },
    ],
  };

  it('should render correctly', () => {
    const { container } = render(
      <DataViewToolbar filters={<DataViewCheckboxFilter {...defaultProps} />} />
    );
    expect(container).toMatchSnapshot();
  });
});
