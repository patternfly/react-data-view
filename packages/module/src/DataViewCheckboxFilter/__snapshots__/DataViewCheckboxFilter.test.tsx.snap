// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataViewCheckboxFilter component should render correctly 1`] = `
<div>
  <div
    class="pf-v5-c-toolbar"
    data-ouia-component-id="DataViewToolbar"
    data-ouia-component-type="PF5/Toolbar"
    data-ouia-safe="true"
    id="pf-random-id-0"
  >
    <div
      class="pf-v5-c-toolbar__content"
    >
      <div
        class="pf-v5-c-toolbar__content-section"
      >
        <div
          class="pf-v5-c-toolbar__item pf-m-search-filter"
        >
          <div
            class="pf-v5-c-toolbar__item"
            data-ouia-component-id="DataViewCheckboxFilter"
          >
            <button
              aria-expanded="false"
              class="pf-v5-c-menu-toggle"
              data-ouia-component-id="DataViewCheckboxFilter-toggle"
              data-ouia-component-type="PF5/MenuToggle"
              data-ouia-safe="true"
              style="width: 200px;"
              type="button"
            >
              <span
                class="pf-v5-c-menu-toggle__icon"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v5-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 512 512"
                  width="1em"
                >
                  <path
                    d="M487.976 0H24.028C2.71 0-8.047 25.866 7.058 40.971L192 225.941V432c0 7.831 3.821 15.17 10.237 19.662l80 55.98C298.02 518.69 320 507.493 320 487.98V225.941l184.947-184.97C520.021 25.896 509.338 0 487.976 0z"
                  />
                </svg>
              </span>
              <span
                class="pf-v5-c-menu-toggle__text"
              >
                Test Checkbox Filter
              </span>
              <span
                class="pf-v5-c-menu-toggle__count"
              >
                <span
                  class="pf-v5-c-badge pf-m-read"
                  data-ouia-component-id="DataViewCheckboxFilter-badge"
                >
                  1
                </span>
              </span>
              <span
                class="pf-v5-c-menu-toggle__controls"
              >
                <span
                  class="pf-v5-c-menu-toggle__toggle-icon"
                >
                  <svg
                    aria-hidden="true"
                    class="pf-v5-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 320 512"
                    width="1em"
                  >
                    <path
                      d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                    />
                  </svg>
                </span>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="pf-v5-c-toolbar__content pf-m-chip-container"
    >
      <div
        class="pf-v5-c-toolbar__group"
      >
        <div
          class="pf-v5-c-toolbar__item pf-m-chip-group"
        >
          <div
            aria-labelledby="pf-random-id-1"
            class="pf-v5-c-chip-group pf-m-category"
            data-ouia-component-type="PF5/ChipGroup"
            data-ouia-safe="true"
            role="group"
          >
            <div
              class="pf-v5-c-chip-group__main"
            >
              <span
                class="pf-v5-c-chip-group__label"
                id="pf-random-id-1"
              >
                Test Checkbox Filter
              </span>
              <ul
                aria-labelledby="pf-random-id-1"
                class="pf-v5-c-chip-group__list"
                role="list"
              >
                <li
                  class="pf-v5-c-chip-group__list-item"
                >
                  <div
                    class="pf-v5-c-chip"
                    data-ouia-component-id="OUIA-Generated-Chip-1"
                    data-ouia-component-type="PF5/Chip"
                    data-ouia-safe="true"
                  >
                    <span
                      class="pf-v5-c-chip__content"
                    >
                      <span
                        class="pf-v5-c-chip__text"
                        id="pf-random-id-2"
                      >
                        Workspace one
                      </span>
                    </span>
                    <span
                      class="pf-v5-c-chip__actions"
                    >
                      <button
                        aria-disabled="false"
                        aria-label="close"
                        aria-labelledby="remove_pf-random-id-2 pf-random-id-2"
                        class="pf-v5-c-button pf-m-plain"
                        data-ouia-component-id="close"
                        data-ouia-component-type="PF5/Button"
                        data-ouia-safe="true"
                        id="remove_pf-random-id-2"
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          class="pf-v5-svg"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          viewBox="0 0 352 512"
                          width="1em"
                        >
                          <path
                            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                          />
                        </svg>
                      </button>
                    </span>
                  </div>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
      <div
        class="pf-v5-c-toolbar__item"
      >
        <button
          aria-disabled="false"
          class="pf-v5-c-button pf-m-link pf-m-inline"
          data-ouia-component-id="DataViewToolbar-clear-all-filters"
          data-ouia-component-type="PF5/Button"
          data-ouia-safe="true"
          type="button"
        >
          Clear filters
        </button>
      </div>
    </div>
  </div>
</div>
`;
